packages/pigeon/lib/ast.dart

```diff
@@ -82,35 +74,101 @@ class Api extends Node {
   }
 }
 
+/// A parameter to a generic entity.  For example, "String" to "List<String>".
+class TypeArgument {
+  /// Constructor for [TypeArgument].
+  TypeArgument({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  /// A string representation of the base datatype.
+  final String dataType;
+
+  /// The type arguments to this [TypeArgument].
+  final List<TypeArgument>? typeArguments;
+
+  /// True if the type is nullable.
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeArgument dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
+/// An entity that represents a typed concept, like a [TypeArgument] or [Field].
+abstract class TypedEntity {
+  /// The data-type of the entity (ex 'String' or 'int').
+  String get dataType;
+
+  /// The type arguments to the entity.
+  List<TypeDeclaration>? get typeArguments;
+
+  /// True if the type is nullable.
+  bool get isNullable;
+}
+
+/// Represents a type declaration.
+class TypeDeclaration implements TypedEntity {
+  /// Constructor for [TypeDeclaration].
+  TypeDeclaration({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  @override
+  final String dataType;
```

Comment:
In this context it should just be called `name` or `baseName` or whatever term you use above; the class is the type.